### PR TITLE
ENH: make guess_datetime_format public

### LINKED ISSUE #54727: ENH: make guess_datetime_format public
### Feature Type

- [ ] Adding new functionality to pandas

- [X] Changing existing functionality in pandas

- [ ] Removing existing functionality in pandas

### Problem Description

Pandas can infer datetime formats when parsing datetimes, but does not provide access to the inferred format via a public API.

### Feature Description

Currently, `guess_datetime_format` is in `pandas.core.tools.datetimes` which is not public. Exporting it from a non-private module would address the issue.

### Alternative Solutions

As far as I can tell, no other Python library provides this requested feature.

### Additional Context

As discussed in https://github.com/dateutil/dateutil/issues/1138

### PR DESCRIPTION
**Pull request by nnlnr**
- [X] closes #54727 (Replace xxxx with the GitHub issue number)
- [X] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [X] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [NA] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [X] Added an entry in the latest `doc/source/whatsnew/v2.1.1.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/whatsnew/v2.1.1.rst [hunk 1]
```diff
@@ -42,7 +42,7 @@ Bug fixes
 
 Other
 ~~~~~
--
+- Enhancement in :meth:`_libs.tslibs.parsing.guess_datetime_format` to make public (:issue:`54727`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.1.rst:45]:
```suggestion
- :meth:`pandas.tseries.guess_datetime_format` is now available (:issue:`54727`)
```

Can you move this to the other enhancement section?

Comment (mroeschke) [reply doc/source/whatsnew/v2.1.1.rst:45]: Also this should be in 2.2.0.rst

**Review by mroeschke — CHANGES_REQUESTED** [review]
Could you add this to the API reference documentation?

File: doc/source/whatsnew/v2.1.1.rst (hunk 1 above)
Comment (nnlnr) [reply doc/source/whatsnew/v2.1.1.rst:45]: WIll do - thanks for the review!

Comment (nnlnr) [reply doc/source/whatsnew/v2.1.1.rst]: moved enhancement note to `2.2.0.rst` unde the **Other Enhancements** section ✅

**Comment by nnlnr** [comment]
> Could you add this to the API reference documentation?
is the API reference documentation in` doc/source/reference/api`? If so, I'm seeing that path in the .gitignore file - is there a procedure to update the API reference under this context?

**Comment by mroeschke** [comment]
I think you can just add an entry to `doc/source/reference/general_functions.rst`

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
thanks for your pr

whilst we're changing this

- in the `guess_Datetime_format` docstring, where it says

>         Warning: dayfirst=True is not strict, but will prefer to parse
        with day first (this is a known bug).

could you write that as a sphinx directive? so `.. warning::`. check the `to_datetime` docs, there's something similar there

- could we add an example to the docstring please? just a simple one-liner of it guessing a datetime format, and another one-liner of it not being able to guess another one, will do

**Comment by nnlnr** [comment]
Thanks for the review and comments @mroeschke and @MarcoGorelli 
- added an entry to doc/source/reference/general_functions.rst ✅ 
- updated warning in docstring as sphinx directive ✅ 
- added examples to docstring ✅

**Comment by MarcoGorelli** [comment]
Can you fix the failing ci checks please

**Comment by nnlnr** [comment]
👋  I'm getting a bit stuck with these failing checks - which seem to be mostly due to the following attribute error:
`AttributeError: module 'pandas' has no attribute 'guess_datetime_format'`

I updated the docstring in parsing.pyx with a sphinx `..warning::` directive and added two examples (learned that formatting was important which was a good lesson); however not sure how to address the failing jobs - any feedback or advice on how to troubleshoot?

File: doc/source/whatsnew/v2.1.1.rst [hunk 2]
```diff
@@ -43,7 +43,7 @@ Bug fixes
 
 Other
 ~~~~~
--
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.1.rst:46]: Can you undo this?

File: pandas/_libs/tslibs/parsing.pyx [hunk 3]
```diff
@@ -874,14 +874,28 @@ def guess_datetime_format(dt_str: str, bint dayfirst=False) -> str | None:
...
     Returns
     -------
     str or None : ret
         datetime format string (for `strftime` or `strptime`),
         or None if it can't be guessed.
+
+    Examples
+    --------
+    >>> # Successful guess
+    >>> # Returns a datetime formatted string
```
Comment (mroeschke) [inline pandas/_libs/tslibs/parsing.pyx:892]:
```suggestion
```

File: pandas/_libs/tslibs/parsing.pyx [hunk 4]
```diff
@@ -874,14 +874,28 @@ def guess_datetime_format(dt_str: str, bint dayfirst=False) -> str | None:
...
+    >>> guess_datetime_format('09/13/2023')
+    '%m/%d/%Y'
+
+    >>> # Unsuccessful guess due to date string with extra or unexpected characters
+    >>> # Returns nothing
```
Comment (mroeschke) [inline pandas/_libs/tslibs/parsing.pyx:897]:
```suggestion
```

File: doc/source/whatsnew/v2.2.0.rst [hunk 5]
```diff
@@ -45,8 +45,8 @@ enhancement2
 
 Other enhancements
 ^^^^^^^^^^^^^^^^^^
+- :meth:`pandas.tseries.api.guess_datetime_format` is now available (:issue:`54727`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.2.0.rst:48]:
```suggestion
- :meth:`pandas.tseries.guess_datetime_format` is now available (:issue:`54727`)
```

File: doc/source/reference/general_functions.rst [hunk 6]
```diff
@@ -58,6 +58,7 @@ Top-level dealing with datetimelike data
    period_range
    timedelta_range
    infer_freq
+   guess_datetime_format
```
Comment (mroeschke) [inline doc/source/reference/general_functions.rst:61]: For this entry specifically I think you'll need a new autosummary section with 

`.. currentmodule:: pandas.tseries.frequencies`

File: doc/source/whatsnew/v2.2.0.rst (hunk 5 above)
Comment (nnlnr) [reply doc/source/whatsnew/v2.2.0.rst:48]: Is this because guess_datetime_format is now a public method - and no longer needs to be called through `api`?

Comment (mroeschke) [reply doc/source/whatsnew/v2.2.0.rst:48]: Oh sorry, I'm actually incorrect. I see you added it to `api`.

File: doc/source/reference/general_functions.rst (hunk 6 above)
Comment (nnlnr) [reply doc/source/reference/general_functions.rst:65]: Tried adding a new autosummary section with 
`.. currentmodule:: pandas.tseries.frequencies`
as follows:
```
Top-level dealing with datetimelike data
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.. autosummary::
   :toctree: api/


   to_datetime
   to_timedelta
   date_range
   bdate_range
   period_range
   timedelta_range
   infer_freq


.. autosummary::
   :currentmodule: pandas.tseries.frequencies


   guess_datetime_format
```

However, I'm still getting hit with the `"AttributeError: module 'pandas' has no attribute 'guess_datetime_format'"` for docstring validation - 

should this be a completely different section with a different header?

**Comment by nnlnr** [comment]
👋 Marco - thanks! That would be great - let me fix up the conflicts and circle back here

**Comment by MarcoGorelli** [comment]
website build looks good:

image

merging then, thanks @nnlnr !

**Review by MarcoGorelli — APPROVED** [review]
going to need @mroeschke to approve before this can be merged

File: doc/source/reference/general_functions.rst [hunk 7]
```diff
@@ -61,6 +61,7 @@ Top-level dealing with datetimelike data
 
 Top-level dealing with Interval data
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
 .. autosummary::
    :toctree: api/
```
Comment (MarcoGorelli) [inline doc/source/reference/general_functions.rst:66]: extra line, but it doesn't seem to make any difference in the build so is OK imo

image

File: doc/source/reference/general_functions.rst [hunk 8]
```diff
@@ -61,6 +61,7 @@ Top-level dealing with datetimelike data
 
 Top-level dealing with Interval data
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
```
Comment (mroeschke) [inline doc/source/reference/general_functions.rst:64]:
```suggestion
```

**Comment by mroeschke** [comment]
Could you also add a whatsnew note in 2.2 under the `Other Enhancements` section?

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @nnlnr  and @MarcoGorelli

**Comment by nnlnr** [comment]
Thanks all 🎉 🙏

**Pull request merged**